draft-ietf-rtcweb-jsep.xml

```diff
@@ -1799,9 +1799,27 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
 
           <t>For media m= sections, JSEP implementations MUST support
           the "UDP/TLS/RTP/SAVPF" profile specified in
-          <xref target="RFC5764"></xref>, and MUST indicate this
-          profile for each media m= line they produce in an offer. For
-          data m= sections, implementations MUST support the
+          <xref target="RFC5764"></xref>. Implementations MUST indicate this
+          profile for each media m= line they produce in an offer unless the
+          media section contains only TCP candidates; if all candidates use TCP as a
```

Comment:
In the IESG review of one of my drafts, Ekr(?) indicated that one can't use "MUST...unless" wording. That triggered some discussions, but there was no outcome. I personally have no problem with it, but just to make sure Ekr is fine with it. One would assume he is, since he is listed as co-author :)

Reply:
What is a "JSEP implementation". A browser implementing the JSEP API, or a JS application using it? I understand if the profile will be included in each offer within the API, but why is it a MUST for a JS application to send it in an offer on the wire?

Reply:
I would suggest that language should be:

In the initial offer or any offer that initiates ICE restart, implementations MUST indicate "UDP/TLS/RTP/SAVPF" profile for each media m= line they produce in an offer.  In the subsequent offers which do not initiate ICE restart, if all candidates for this m= section use TCP as a transport,  implementations MUST indicate "TCP/DTLS/RTP/SAVPF," to allow for compatibility with non-WebRTC implementations of <xref target="I-D.ietf-mmusic-ice-sip-sdp"></xref>.

Reply:
Roman is correct but this is NUTS. We are doing all of this so we can support middle boxes tampering with our signalling in a way that we know won't work? Seriously ?  Note all of this is fine with SBCs as they are actually in the signalling and can signal if they do ICE or not, but we are doing this for ALGs which pretty much no one uses, are a really bad idea, are not recommended by the IETF - how about we get rid of all the ALG stuff and replace it with SDP MUST be transported over a protocol that provides integrity.

Reply:
Roman - Thanks for walking me through this. I have a slight preference for just get rid of all this stuff when using ICE but I could easily live with what you proposed and I agree it is easy to implement.

Reply:
As a side note, the fact that it is not specified here what is supposed to be done creates some degree of confusion. For instance in this situation (subsequent offer when ICE restart is not initiated and only TCP candidate is present) Chrome currently adds a UDP candidate in addition to TCP candidate and puts this UDP candidate in the c=/m= line. This is clearly not what was intended but this is how the current text was interpreted.

Reply:
@nils-ohlmeier We are talking about the situation when ICE nomination process is complete and TCP candidate is nominated. According to RFC 5245 Section 9.1.2.2 (https://tools.ietf.org/html/rfc5245#section-9.1.2.2):

The agent MUST include candidate attributes for candidates matching the default destination for each component of the media stream, and MUST NOT include any other candidates.

This text is still present in ice-sip-sdp Section 3.4.1.2.2 (https://tools.ietf.org/html/draft-ietf-mmusic-ice-sip-sdp-24#section-3.4.1.2.2):

Once a candidate pair has been nominated for a data stream, the IP address, port and transport in each "c=" and "m=" line associated with that data stream MUST match the data associated with the nominated pair for that data stream.  In addition, the offerer only includes SDP candidates representing the local candidates of the nominated candidate pair.  The offerer MUST NOT include any other SDP candidate attributes in the subsequent offer.

Based on this, if ICE restart is not initiated, ICE nomination is complete, and TCP candidate is nominated, offer should include one currently nominated TCP candidate and no UDP candidates. In this case, TCP/DTLS/RTP/SAVPF protocol and the port of the TCP candidate should be used in the m= line with address of the TCP candidate used in the c= line.

There should be no other scenarios when TCP/DTLS/RTP/SAVPF is generated in the offer by the browser.

In the answer, the protocol generated by the browser should match the protocol in the offer it have received for this m= line as long as it is one of the protocols listed later in this section.

Reply:
I ask again: what is meant by "implementation"?

Are we suggesting that a JS application using the JSEP API is required to use "UDP/TLS/RTP/SAVPF"?

Reply:
@juberti I will try to answer multiple comments in this single one.

When a browser that has an active TCP candidate connects to a MCU with a passive TCP candidate, then the browser TCP candidate is 1.2.3.4:9 - i.e., the discard port and end point puts '9' into the m= line.

According to @adamroach in https://mailarchive.ietf.org/arch/msg/mmusic/Btq1iVG_BrrvshTps1toyzyWWRs , Chrome will add UDP attribute to an offer even if ICE nomination was completed and TCP candidate was nominated. I have not verified this.

During the discussions at MMUSIC we decided that when initial or offer with ICE restart is initiated, the capabilities of the remote end point are not always known. In case of initial offer little is known about remote and ICE restart can be initiated to allow for 3rd party call control, which can change the remote end point. Because of this, since in both cases remote end point might not support ICE TCP or TCP based proto, it is safer to use UDP based proto in the offer, even when TCP candidate pair is currently being used. It was also decided that it is cleaner not to change proto during ICE nomination process, since nominated pair can change during the signaling exchange. This resulted in always using UDP for initial offer, during ICE restart or when ICE nomination is still in progress and using TCP only when ICE nomination is complete and TCP candidate was nominated.